From f7cf7738ed40800404e5e5c374c24d45b8370472 Mon Sep 17 00:00:00 2001
From: Hanna Hawa <hannah@marvell.com>
Date: Mon, 13 Mar 2017 12:56:29 +0200
Subject: [PATCH 1464/2241] fix: dma: mv_xor_v2: add lock while free the
 completed descriptors

Add lock to prevent mutual usage of free descriptor list by more than
one CPU.

Change-Id: Ia686b616dacfe9c9e39dadd4e58628063824ec06
Signed-off-by: Hanna Hawa <hannah@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/37415
Reviewed-by: Omri Itach <omrii@marvell.com>
Tested-by: Omri Itach <omrii@marvell.com>
---
 drivers/dma/mv_xor_v2.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/drivers/dma/mv_xor_v2.c b/drivers/dma/mv_xor_v2.c
index ca340f8..fa88be3 100644
--- a/drivers/dma/mv_xor_v2.c
+++ b/drivers/dma/mv_xor_v2.c
@@ -618,13 +618,16 @@ static void mv_xor_v2_tasklet(unsigned long data)
 
 	dev_dbg(xor_dev->dmadev.dev, "%s %d\n", __func__, __LINE__);
 
+	/* Lock the channel */
+	spin_lock_bh(&xor_dev->sw_ll_lock);
 	/* free the compeleted descriptors (check the ctrl ack flag) */
 	list_for_each_entry_safe(iter, _iter, &xor_dev->complete_sw_desc,
 				 node) {
-
 		if (async_tx_test_ack(&iter->async_tx))
 			list_move_tail(&iter->node, &xor_dev->free_sw_desc);
 	}
+	/* Release the channel */
+	spin_unlock_bh(&xor_dev->sw_ll_lock);
 
 	/* get thepending descriptors parameters */
 	num_of_pending = mv_xor_v2_get_pending_params(xor_dev, &pending_ptr);
-- 
2.7.4

